[yeoshuheng] iP

yeoshuheng opened the pull request:
# shuheng.bot

> Shu Heng is a nerd - [Haters](https://www.youtube.com/watch?v=562ngVq9EQM)

This is a project template for a greenfield Java project. 
It's named after the *top student* in NUS CS (~~Allegedly~~)! :bowtie:

Why use `shuheng`?

* **Free!**
* **Easy** to learn!
* **Easy** to use too!

## How do I run it?

1. Download the jar file!
2. Run it!
3. Add your tasks...
4. And let `shuheng` handle everything else!

```bash
java --jar shuheng.jar
```

Features:

- [x] Support deadlines, todos, events
- [x] Add UI
- [ ] Add CI/CD pipeline

yeoshuheng committed:
feat: renamed bot + added greet function

yeoshuheng committed:
Merge pull request #1 from yeoshuheng/week1

feat: renamed bot + added greet function

yeoshuheng committed:
feat: echo added

yeoshuheng committed:
feat: added to-do list feature

yeoshuheng committed:
feat: added ability to mark classes

yeoshuheng committed:
Merge pull request #2 from yeoshuheng/week1

Week1

yeoshuheng committed:
refactor: task kept abstract

yeoshuheng committed:
feat: integrated todo, deadlines and events into duke

yeoshuheng committed:
fix: error in time data output

yeoshuheng committed:
fix: removed spacings and improved consistency in printed outputs

yeoshuheng committed:
feat: added automated ui testing

yeoshuheng committed:
feat: added input checks

yeoshuheng committed:
feat: added more exceptions to handle invalid index

yeoshuheng committed:
feat: delete function added

yeoshuheng committed:
refactor: uses enums instead; fix: bug in check index function

yeoshuheng committed:
Merge pull request #3 from yeoshuheng/week1

Week 1 content

yeoshuheng committed:
feat: added logging functionality

yeoshuheng committed:
fix: null issue in logger

yeoshuheng committed:
feat: added restoration capabilities

yeoshuheng committed:
feat: handle corrupted files

yeoshuheng committed:
fix: bug in filepath and equality check for isDone during restoration

yeoshuheng committed:
feat: integrated logging feature into chatbot

yeoshuheng committed:
fix: no working dir bug

yeoshuheng committed:
Merge pull request #4 from yeoshuheng/branch-Level-7

Branch level 7

yeoshuheng committed:
feat: added parser + exception for datetime

yeoshuheng committed:
feat: added datetime field + functions in tasks

yeoshuheng committed:
feat: integrated datetime into filemanager & duke

yeoshuheng committed:
fix: revert to single valid input for date-time

yeoshuheng committed:
Merge pull request #5 from yeoshuheng/branch-Level-8

Branch level 8

yeoshuheng committed:
refactor: added storage class

yeoshuheng committed:
refactor: extracted parser into class

yeoshuheng committed:
refactor: extracted task list into class

yeoshuheng committed:
chore: updated gitignore

yeoshuheng committed:
refactor: extracted UI into another class

yeoshuheng committed:
Merge pull request #6 from yeoshuheng/A-MoreOOP

A more oop

yeoshuheng committed:
refactor: remove template logs

yeoshuheng committed:
refactor: changed package directory

yeoshuheng committed:
Merge pull request #7 from yeoshuheng/A-Packages

refactor: changed package directory

yeoshuheng committed:
fix: unhandled invalid command

yeoshuheng committed:
perf: added gradle support

yeoshuheng committed:
Merge pull request #8 from yeoshuheng/A-Gradle

perf: added gradle support

yeoshuheng committed:
test: added junit test for parser and task list

yeoshuheng committed:
Merge pull request #9 from yeoshuheng/A-JUnit

test: added junit test for parser and task list

yeoshuheng committed:
perf: added shadowjar to gradle

yeoshuheng committed:
Merge pull request #10 from yeoshuheng/A-Jar

perf: added shadowjar to gradle

yeoshuheng committed:
docs: added javadocs for front-facing classes

yeoshuheng committed:
feat: added search feature to task list

yeoshuheng committed:
test: added enhanced test cases for parser

yeoshuheng committed:
Merge pull request #11 from yeoshuheng/add-test

test: added enhanced test cases for parser

yeoshuheng committed:
feat: integrated find function into UI

yeoshuheng committed:
style: fix for checkstyle

yeoshuheng committed:
style: fix test for checkstyle

yeoshuheng committed:
Merge branch 'master' into branch-A-CodingStandard

yeoshuheng committed:
Merge pull request #12 from yeoshuheng/branch-A-CodingStandard

Branch a coding standard

yeoshuheng committed:
style: fix checkstyle for updated parser

yeoshuheng committed:
Merge branch 'master' into branch-A-JavaDoc

yeoshuheng committed:
Merge pull request #13 from yeoshuheng/branch-A-JavaDoc

docs: added javadocs for front-facing classes

yeoshuheng committed:
style: fix stylecheck issues for javadocs

yeoshuheng committed:
styles: fix checkstyle for extended test classes

yeoshuheng committed:
Merge branch 'master' into branch-Level-9

yeoshuheng committed:
Merge pull request #14 from yeoshuheng/branch-Level-9

Branch level 9

yeoshuheng committed:
style: fix checkstyle for level9 components

yeoshuheng committed:
fix: add back find features deleted in merge

yeoshuheng committed:
Merge pull request #15 from yeoshuheng/branch-Level-9

fix: add back find features deleted in merge

yeoshuheng committed:
fix: fix post-merge listkeywords feat

yeoshuheng committed:
docs: add more docs for storage & tasklist

yeoshuheng committed:
Merge pull request #16 from yeoshuheng/branch-A-CodingStandard

docs: add more docs for storage & tasklist

yeoshuheng committed:
docs: Fix docs to meet checkstyle conventions

yeoshuheng committed:
Merge pull request #17 from yeoshuheng/branch-A-CheckStyle

docs: Fix docs to meet checkstyle conventions

yeoshuheng committed:
perf: Add javafx to gradle

yeoshuheng committed:
feat: add gui

yeoshuheng committed:
feat: Connect frontend to input handlers

yeoshuheng committed:
feat: Add inverted dialog boxes for shuheng & user

yeoshuheng committed:
perf: Change old javaFX components to FXML

yeoshuheng committed:
fix: Fix UI not loading properly

yeoshuheng committed:
Merge pull request #18 from yeoshuheng/branch-Level-10

Branch level 10

yeoshuheng committed:
docs: Update README

yeoshuheng committed:
perf: Update parser to use varargs

yeoshuheng committed:
Merge pull request #19 from yeoshuheng/branch-A-Varargs

perf: Update parser to use varargs

yeoshuheng committed:
perf: Replace iterative code with streams

yeoshuheng committed:
perf: Replace iterative code with streams

yeoshuheng committed:
perf: Add assertions to parser

yeoshuheng committed:
perf: Add assertions to parser

yeoshuheng committed:
perf: Add CI/CD

yeoshuheng committed:
Merge pull request #20 from yeoshuheng/branch-A-CI

perf: Add CI/CD

0liverkong commented on src/main/java/duke/Storage.java:
```diff
@@ -0,0 +1,135 @@
+            bw.write(sb.toString());
+            bw.close();
+        } catch (IOException e) {
+            System.out.println("Problem writing to log! " + e.getMessage());
```
Im not sure if its preferred, but would it be better to let UI.java handle the printing of error statements too?

0liverkong commented on src/main/java/duke/Storage.java:
```diff
@@ -0,0 +1,135 @@
+    public void createLog() {
+        File logFile = new File(this.path);
+        try {
+            boolean response = logFile.createNewFile();
```
I think it'd be better to rename this variable to indicate it is a boolean

0liverkong commented on src/main/java/duke/UI.java:
```diff
@@ -0,0 +1,174 @@
+    /**
+     * This enum represents a UI command.
+     */
+    public enum Command {
```
would it be better to make this (Command enum) a class by itself? right now run() calls parser.getCommand() which refers back to this file, a little wonky

0liverkong commented on src/main/java/duke/tasks/Events.java:
```diff
@@ -0,0 +1,50 @@
+/**
+ * This class represents a event task.
+ */
+public class Events extends Task {
```
should rename this to just Event

0liverkong commented on src/main/java/duke/tasks/ToDos.java:
```diff
@@ -0,0 +1,29 @@
+/**
+ * This class represents a task without a deadline.
+ */
+public class ToDos extends Task {
```
same here, would be good to rename this to ToDo

0liverkong commented on src/main/java/duke/UI.java:
```diff
@@ -0,0 +1,174 @@
+        this.manager = manager;
+        this.parser = parser;
+        this.sc = new Scanner(System.in);
+        sayHi();
```
the result of sayHi() gets discarded here - does your GUI say hi?

0liverkong commented on src/main/java/duke/exceptions/DukeException.java:
```diff
@@ -0,0 +1,10 @@
+package duke.exceptions;
```
really like that you coded out classes for all the different exceptions. good job!

0liverkong commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,161 @@
+     * @return A LocalDateTime version of the input.
+     * @throws InvalidDateTimeException Thrown if error in conversion process.
+     */
+    public static LocalDateTime parseDate(String potentialDate) throws InvalidDateTimeException {
```
you can use LocalDateTime.parse() to simplify the code here.

0liverkong commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,42 @@
+        this.manager.createLog();
+        this.parser = new Parser();
+        this.history = new TaskList(this.manager);
+        this.ui = new UI(manager, parser, history);
```
I think it is a little weird that both Ui.java and Duke.java maintain references to manager, parser and history. is that entirely necessary?

gavingoh99 commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,42 @@
+     * Getter for UI.
+     * @return UI for the chatbot.
+     */
+    public UI getUI() {
```
Perhaps we could write our method to more closely follow the coding standards!
```suggestion
    public UI getUi() {
```

gavingoh99 commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,161 @@
+            ret[0] = searchDesc;
+            break;
+        case "event":
+            Integer startIdx = -1;
+            Integer endIdx = -1;
```
Perhaps naming the variables more intuitively can reduce the chances of future confusion!
```suggestion
            Integer startIndex = -1;
            Integer endIndex = -1;
```

gavingoh99 commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,161 @@
+     */
+    public static String[] extractDescriptionData(String... descriptionArray) throws
+            InvalidInputException {
+        String[] ret = new String[3];
```
```suggestion
        String[] returnArray = new String[3];
```

gavingoh99 commented on src/main/java/duke/Storage.java:
```diff
@@ -0,0 +1,135 @@
+     * @param path The location where the logs will be written to.
+     */
+    public Storage(String path) {
+        String wd = System.getProperty("user.dir");
```
```suggestion
        String workingDirectory = System.getProperty("user.dir");
```

gavingoh99 commented on src/main/java/duke/tasks/Deadlines.java:
```diff
@@ -0,0 +1,41 @@
+/**
+ * This class represents a task with a deadline.
+ */
+public class Deadlines extends Task {
```
Perhaps this class and all other classes that inherit from Task should be phrased in singular form?
```suggestion
public class Deadline extends Task {
```

gavingoh99 commented on src/test/java/duke/TaskListTest.java:
```diff
@@ -0,0 +1,20 @@
+        TaskList taskList = new TaskList();
+        ToDos toAdd = new ToDos("do stuff");
+        ToDos toAdd2 = new ToDos("do more stuff");
+        taskList.addTaskTest(toAdd);
```
I like your use of a stub here!

gavingoh99 reviewed (commented):
LGTM! Overall your code is very detailed and goes above and beyond the expected progress of the iP! Just a few nits here and there that pertain to how to better adhere to the coding standards of CS2103T! I look forward to your final product!

yeoshuheng committed:
fix: Modify variables to better fit code quality

yeoshuheng committed:
fix: Modify variables to better fit code quality

yeoshuheng committed:
fix: Modify variables to better fit code quality

yeoshuheng committed:
fix: Modify variables to better fit code quality

yeoshuheng committed:
Merge pull request #22 from yeoshuheng/branch-A-Assertions

perf: Add assertions to parser

yeoshuheng committed:
Merge pull request #21 from yeoshuheng/branch-A-Streams

perf: Replace iterative code with streams

yeoshuheng committed:
perf: SayHi() command no longer needed. Including unneeded code is a bad practice and should be avoided. Let's, remove sayHi()

yeoshuheng committed:
perf: Removed unneeded booleans

yeoshuheng committed:
perf: Removed unneeded booleans

yeoshuheng committed:
Merge branch 'master' into branch-A-CodeQuality

yeoshuheng committed:
feat: Add priority levels to task

yeoshuheng committed:
test: Add test to consider priority

yeoshuheng committed:
feat: Integrate priority into UI

yeoshuheng committed:
test: Fix checkstyle issues in test

yeoshuheng committed:
chore: Remove unneeded files.

jimseah-0116 commented on src/main/java/duke/Ui.java:
```diff
@@ -0,0 +1,167 @@
+     * This enum represents a UI command.
+     */
+    public enum Command {
+        BYE,
+        LIST,
+        EVENT,
+        TODO,
+        DEADLINE,
+        MARK,
+        UNMARK,
+        DELETE,
+        FIND,
```
Constants are in all uppercase letters. Great!

jimseah-0116 commented on src/main/java/duke/Ui.java:
```diff
@@ -0,0 +1,167 @@
+     *
+     * @return The output to print.
+     */
+    public String sayBye() {
```
Method name follows the correct convention (the action is a verb). Great!

jimseah-0116 commented on src/main/java/duke/exceptions/InvalidTaskException.java:
```diff
@@ -0,0 +1,14 @@
+/**
+ * This class represents the error thrown when an invalid input is detected.
+ */
+public class InvalidTaskException extends InvalidInputException {
```
Fantastic job on making separate exceptions for each case!

jimseah-0116 commented on src/main/java/duke/frontend/DialogBox.java:
```diff
@@ -0,0 +1,76 @@
+/**
+ * This class represents a dialog box.
+ */
+public class DialogBox extends HBox {
```
Class name follows Pascalcase, great!

jimseah-0116 commented on src/main/java/duke/frontend/MainWindow.java:
```diff
@@ -0,0 +1,79 @@
+ */
+public class MainWindow extends AnchorPane {
+    @FXML
+    private ScrollPane scrollPane;
+    @FXML
+    private VBox dialogContainer;
+    @FXML
+    private TextField userInput;
+    @FXML
+    private Button sendButton;
+
+    private boolean isExit = false;
```
Variable names are in camelCase, follows the naming conventions well. Boolean variable is also named to sound like a boolean. Great!

jimseah-0116 commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,208 @@
+     * @return A correct command enum assigned to the array.
+     * @throws InvalidTaskException When the command does not match any of the enums.
+     */
+    public static Ui.Command getCommand(String[] arr) throws InvalidTaskException {
```
Method modifier follows the correct given order, great!

jimseah-0116 reviewed (commented):
Overall very well done, packages are all done well with related classes being put in the same package and code is easily readable. Wow! Very inspiring!

yeoshuheng committed:
fix: Fix print of hi and by to GUI

yeoshuheng committed:
fix: Fix checkstyle problems

yeoshuheng committed:
fix: Change wrong commit message

yeoshuheng committed:
feat: Resolve conflicts

yeoshuheng committed:
fix: Overwrite old commit via rebase

yeoshuheng committed:
Merge branch 'master' into branch-A-CodeQuality

yeoshuheng committed:
Merge pull request #25 from yeoshuheng/branch-A-CodeQuality

Branch a code quality

yeoshuheng committed:
feat: Update assert statements for new array

yeoshuheng committed:
Merge branch 'master' into branch-A-Assertions

yeoshuheng committed:
fix: Fix merge conflicts

yeoshuheng committed:
Merge pull request #26 from yeoshuheng/branch-A-Assertions

feat: Update assert statements for new array

yeoshuheng committed:
refactor: Update names to shuheng

yeoshuheng committed:
feat: Add personality

yeoshuheng committed:
Merge pull request #27 from yeoshuheng/branch-A-Personality

Branch a personality

yeoshuheng committed:
fix: Add date checks

User can add events with end earlier than start.

Such events do not make sense and should not be allowed.

Let's,
* Add checks to ensure end is later than start.

yeoshuheng committed:
Merge pull request #28 from yeoshuheng/branch-A-MoreErrorHandling

fix: Add date checks

yeoshuheng committed:
refactor: Improve adherence to SLAP

yeoshuheng committed:
docs: Add readme

yeoshuheng committed:
docs: Add userguide

yeoshuheng committed:
refactor: Improve adherence to SLAP

yeoshuheng committed:
docs: Update ui img

yeoshuheng committed:
Merge pull request #30 from yeoshuheng/branch-A-UserGuide

Branch a user guide

yeoshuheng committed:
refactor: Convert names to shuheng

yeoshuheng committed:
fix: Update exception msg

yeoshuheng committed:
fix: Ui bounds

yeoshuheng committed:
docs: Add new Ui png

yeoshuheng committed:
docs: Add new Ui png

yeoshuheng committed:
docs: Update Ui png
